### ISSUE TITLE
feature request: expose elligator2

### ISSUE THREAD

**Issue by James-E-A** [jetperch/pymonocypher#5]
Would it be feasible to surface [these bindings](https://github.com/jetperch/pymonocypher/blob/3a49d68e7657edb9c0770e7288263b4a28e7189a/c_monocypher.pyx#L217), too?

Monocypher is pretty much [the only](https://loup-vaillant.fr/articles/implementing-elligator) source \([as far as I could tell](https://pypi.org/search/?q=elligator)\) of Elligator2 in the wild, especially in Python.

**Comment by mliberty1** [comment]
Of course.  I am not familiar with Elligator2.  Is simply exposing the functions with length & type checks on the inputs be enough?

```C
# Elligator mappings proper
void crypto_hidden_to_curve(uint8_t curve [32], const uint8_t hidden[32])
int  crypto_curve_to_hidden(uint8_t hidden[32], const uint8_t curve [32], uint8_t tweak)
    
# Easy to use key pair generation
void crypto_hidden_key_pair(uint8_t hidden[32], uint8_t secret_key[32], uint8_t seed[32])
```

However, it looks like the new [Monocypher 4.0](https://monocypher.org/manual/elligator) [[source](https://github.com/LoupVaillant/Monocypher/blob/ea79193643274dbf75c09ca8ec2d60231d0d21b3/src/monocypher.h#L309)] (pymonocypher is still on 3.1.3) changes the names:

```C
void crypto_elligator_map(uint8_t curve[32], const uint8_t hidden[32]);
int crypto_elligator_rev(uint8_t hidden[32], const uint8_t curve[32], uint8_t tweak);
void crypto_elligator_key_pair(uint8_t hidden[32], uint8_t secret_key[32], uint8_t seed[32]);
```

So, maybe we just start out python with the new 4.0 names.  What do you think of this API?

```python
def elligator_map(hidden: bytes) -> bytes
def elligator_rev(curve: bytes, tweak) -> bytes, raises RuntimeError on nonzero return code
def elligator_key_pair(seed: bytes) -> (hidden: bytes, secret: bytes)
```

**Comment by James-E-A** [comment]
> So, maybe we just start out python with the new 4.0 names. What do you think of this API?

That sounds like a good idea, although I have two thoughts:

1\. Maybe raise return value `255` into `ValueError("curve point is unsuitable for hiding")`, for semantic accuracy?

> … this operation _may_ fail because not all curve points are capable of being hidden. In this case, `crypto_elligator_rev()` must be tried again with a new key pair, though tweak does not need to be changed.

2\. Make the second argument optional, calling `urandom(1)` when it's not provided?

> The tweak **must** be chosen at random … A 1-byte random number

**Comment by mliberty1** [comment]
Hi @JamesTheAwesomeDude - I just created [pull request 5](https://github.com/jetperch/pymonocypher/pull/6) to address this issue.  Could you take a look and let me know what you think?

**Comment by rubidium42** [comment]
Shouldn't this request be closed since #6 got merged?
